platforms/jvm/code-quality/src/main/groovy/org/gradle/api/plugins/quality/AbstractCodeQualityTask.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins.quality;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.JavaVersion;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.MapProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.SourceTask;
+import org.gradle.api.tasks.VerificationTask;
+import org.gradle.jvm.toolchain.JavaLauncher;
+import org.gradle.jvm.toolchain.JavaToolchainService;
+import org.gradle.jvm.toolchain.internal.CurrentJvmToolchainSpec;
+import org.gradle.process.JavaForkOptions;
+import org.gradle.work.DisableCachingByDefault;
+import org.gradle.workers.WorkerExecutor;
+
+import javax.inject.Inject;
+
+/**
+ * Base class for code quality tasks.
+ *
+ * @since 8.4
+ */
+@Incubating
+@DisableCachingByDefault(because = "Super-class, not to be instantiated directly")
+abstract public class AbstractCodeQualityTask extends SourceTask implements VerificationTask {
+    private static final String OPEN_MODULES_ARG = "java.prefs/java.util.prefs=ALL-UNNAMED";
+    private final Property<JavaLauncher> javaLauncher;
+    private final Property<String> minHeapSize;
+    private final Property<String> maxHeapSize;
+    // TODO - Convert VerificationTask to use property-based API and deprecate boolean methods
+    private final Property<Boolean> ignoreFailures;
+    private final MapProperty<String, String> systemProperties;
+    private final MapProperty<String, String> environment;
+
+    @Inject
+    public AbstractCodeQualityTask() {
+        this.javaLauncher = configureFromCurrentJvmLauncher(getToolchainService(), getObjectFactory());
+        this.minHeapSize = getObjectFactory().property(String.class);
+        this.maxHeapSize = getObjectFactory().property(String.class);
+        this.ignoreFailures = getObjectFactory().property(Boolean.class).convention(false);
+        this.systemProperties = getObjectFactory().mapProperty(String.class, String.class);
+        this.environment = getObjectFactory().mapProperty(String.class, String.class);
+    }
+
+    /**
+     * JavaLauncher for toolchain support
+     */
+    @Nested
+    public Property<JavaLauncher> getJavaLauncher() {
+        return javaLauncher;
+    }
+
+    /**
+     * The minimum heap size for the worker process, if any.
+     * Supports the units megabytes (e.g. "512m") and gigabytes (e.g. "1g").
+     *
+     * @return The minimum heap size. Value should be null if the default minimum heap size should be used.
+     */
+    @Optional
```

Comment:
I'm not 100% sure, but I think `@Optional` on a Property might be meaningless because it only checks for null and not unset Properties.